src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs

```diff
@@ -432,24 +444,39 @@ internal static FunctionSignatureWithReturnType FindBestFunctionSignature(Functi
 
                     if (betterThanAllOthers)
                     {
-                        if (bestFunctionIndex == -1)
-                        {
-                            bestFunctionIndex = i;
-                        }
-                        else
-                        {
-                            // This means there were at least 2 equally matching functions.
-                            return null;
-                        }
+                        // cache equally matching functions for case-sensitive resolution based on null/non-null function call token.
+                        equallyArgumentsMatchingNameFunctions.Add(applicableNameFunctions[i]);
                     }
                 }
 
-                if (bestFunctionIndex == -1)
+                KeyValuePair<string, FunctionSignatureWithReturnType> result;
+                if (equallyArgumentsMatchingNameFunctions.Count == 0)
+                {
+                    // Best match count = 0.
+                    result = NotFoundKeyValuePair;
+                }
+                else if (equallyArgumentsMatchingNameFunctions.Count == 1)
                 {
-                    return null;
+                    // Best match count = 1.
+                    result = equallyArgumentsMatchingNameFunctions[0];
+                }
+                else
+                {
+                    // For multiple best matches based on argument types
+                    if (equallyArgumentsMatchingNameFunctions.Count(
+                            _ => _.Key.Equals(functionCallToken, StringComparison.Ordinal)) == 1)
+                    {
+                        // There is only one case-sensitive function name match among the equally argument-types matching functions.
+                        result = equallyArgumentsMatchingNameFunctions.First(_ => _.Key.Equals(functionCallToken, StringComparison.Ordinal));
+                    }

```

Comment:
You currently loop through the list twice; once to find if there is exactly one match, and once to return it if there is.  Would be more efficient to rewrite L452-L477 as something like:

                KeyValuePair<string, FunctionSignatureWithReturnType> result = NotFoundKeyValuePair;
                if (equallyArgumentsMatchingNameFunctions.Count == 1)
                {
                    // Best match count = 1.
                    result = equallyArgumentsMatchingNameFunctions[0];
                }
                else
                {
                    // For multiple best matches based on argument types
                    foreach (KeyValuePair<string,FunctionSignatureWithReturnType> candidate in equallyArgumentsMatchingNameFunctions)
                    {
                        if (candidate.Key.Equals(functionCallToken, StringComparison.Ordinal))
                        {
                            result = candidate;
                            break;
                        }
                    }
                }

note that this also handles the Count == 0 case.

Reply:
This is a tricky case. In the case of multiple case-sensitive best matches, we should return invalid resolution ( = NotFoundKeyValuePair) instead of returning the first match.
The computational complexity should be the same O(N), and linear should be the minimum in this case.

---
In reply to: [190104069](https://github.com/OData/odata.net/pull/1156#discussion_r190104069) [](ancestors = 190104069)

Reply:
In what scenario would we have multiple case-sensitive best matches?

In order to handle multiple case-sensitive best matches, could you remove the "break" in the suggested code above and throw your exception if result!=NotFoundKeyValuePair prior to setting result=candidate? That would still be more efficient than looping through the list twice.

---
In reply to: [190715400](https://github.com/OData/odata.net/pull/1156#discussion_r190715400) [](ancestors = 190715400,190104069)

Reply:
Sorry the ambiguous result case should be for the multiple case insensitive match, not case-sensitive matches.
Given that there can be at most one case-sensitive match candidate, the multiple matches resolution can be simpler.

---
In reply to: [191602830](https://github.com/OData/odata.net/pull/1156#discussion_r191602830) [](ancestors = 191602830,190715400,190104069)